### PR TITLE
sambacc: complete type annotations

### PR DESCRIPTION
**Pull request by phlogistonjohn**
The sambacc dir is basically a library and we want to ensure high quality and good documentation so configure mypy to require
type annotations on functions defined in sambacc. Update code in sambacc to add type annotations to any functions lacking them.

The sambacc.commands on the other hand are mean to be only run as part of our CLI tools so we'll be a bit less strict there.

### CODE REVIEW

File: sambacc/container_dns.py [hunk 1]
```diff
@@ -27,18 +29,22 @@
...
 class HostState:
-    def __init__(self, ref="", items=[]):
+    T = typing.TypeVar("T", bound="HostState")
+
+    def __init__(
+        self, ref: str = "", items: typing.Optional[list[HostInfo]] = None
+    ) -> None:
         self.ref = ref
-        self.items = items
+        self.items = items or []
```
Comment (synarete) [inline sambacc/container_dns.py:38]: Consider having (align with other member-initialization in this PR):
```
       self.ref: str = ref
       self.items: list[HostInfo] = items or []
```

File: sambacc/inotify_waiter.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def _get_events(self):
             and ((event.mask & _inotify.flags.CLOSE_WRITE) != 0)
         ]
 
-    def _wait(self):
+    def _wait(self) -> typing.Iterator[None]:
```
Comment (synarete) [inline sambacc/inotify_waiter.py:88]: Why no just have a single `wait` function? What is the advantage of using `next(typing.Iterator[None])` ?

**Review by synarete — COMMENTED** [review]
Two minor comments. Overall, LGTM.

**Comment by phlogistonjohn** [comment]
Wow. Somehow I managed to forget about this PR. I will need to revisit this soon!

File: sambacc/container_dns.py (hunk 1 above)
Comment (phlogistonjohn) [reply sambacc/container_dns.py]: updated.

File: sambacc/inotify_waiter.py (hunk 2 above)
Comment (phlogistonjohn) [reply sambacc/inotify_waiter.py:88]: At the time I must have felt implementing the function in terms of an iterator was easier or clearer. This PR only adds the proper type annotation to the function, it doesn't change the runtime behavior. Let's leave that for another time.

**Comment by phlogistonjohn** [comment]
I finally updated this PR. wow.
@synarete @spuiuk Please revisit. I'd appreciate reviews within a week's time, so I'm putting on a virtual deadline of 2022-11-15. thanks!

File: sambacc/samba_cmds.py [hunk 3]
```diff
@@ -159,19 +159,19 @@ def __repr__(self) -> str:
 samba_dc = SambaCommand("/usr/sbin/samba")
 
 
-def smbd_foreground():
+def smbd_foreground() -> SambaCommand:
     return smbd[
         "--foreground", _daemon_stdout_opt("smbd"), "--no-process-group"
     ]
```
Comment (synarete) [inline sambacc/samba_cmds.py:165]: IMHO, now that we have explicit return type, those lines look a bit enigmatic. I would consider (in a separate PR) to have more explicit code (also in other places below):

```
def smbd_foreground() -> SambaCommand:
    return SambaCommand("/usr/sbin/smbd",
        [ "--foreground", _daemon_stdout_opt("smbd"), "--no-process-group" ])
```

**Review by synarete — APPROVED** [review]
LGTM

**Pull request merged**